File: 003.前端代码/backend/service/smsProvider.js
/**
 * 短信通道抽象。
 *
 * 与 Java 后端 {@code com.xingzhewk.service.sms.SmsProvider} 接口对齐。
 * 详见 contracts/DIFFS.md #2。
 *
 * dev 默认 {@link LogSmsProvider}，prod 配 SMS_PROVIDER=noop 改用 {@link NoopSmsProvider}。
 */

/** 把验证码写到应用日志。开发者自己看 console 拿码。 */
class LogSmsProvider {
    send(phone, code) {
        // WARN 是故意的：任何看日志的人应该一眼意识到这是测试通道
        console.warn(`[SMS-LOG-PROVIDER · DEV-ONLY] phone=${phone} code=${code} —— 这是测试通道，prod 必须切换 app.sms.provider`);
        return true;
    }
    name() { return 'log'; }
}

/** 拒绝发送的通道。prod 未接真通道时的安全默认。 */
class NoopSmsProvider {
    send(phone, code) {
        console.info(`[SMS-NOOP-PROVIDER] 拒绝发送 phone=${phone}（短信通道未接入）`);
        return false;
    }
    name() { return 'noop'; }
}

/** 按 SMS_PROVIDER 环境变量选 provider。 */
function createSmsProvider() {
    const which = (process.env.SMS_PROVIDER || 'log').toLowerCase();
    switch (which) {
        case 'log':   return new LogSmsProvider();
        case 'noop':  return new NoopSmsProvider();
        // 未来 'aliyun' / 'tencent' 在此加
        default:
            console.warn(`[sms] 未知 SMS_PROVIDER="${which}"，降级到 log`);
            return new LogSmsProvider();
    }
}

module.exports = { LogSmsProvider, NoopSmsProvider, createSmsProvider };

File: 005.后端工程师 (java工程师)/expense-tracker-server/src/main/java/com/xingzhewk/service/sms/SmsProvider.java
package com.xingzhewk.service.sms;

/**
 * 短信发送通道。
 *
 * 抽出来是为了把「业务逻辑（生成码 + 存储 + 频控）」和「具体怎么把码送到用户手机」分开：
 *   - dev / 个人项目阶段：{@code LogSmsProvider} 写日志，开发者自己看 console；
 *   - 接入阿里云/腾讯云：新增 AliyunSmsProvider 实现，业务代码一行不改；
 *   - prod 未接通道：{@code NoopSmsProvider} 直接拒绝，避免「假装能发」的安全洞。
 *
 * 注意：实现方不应该把 code 返回给前端 —— 这是验证码语义的底线。
 */
public interface SmsProvider {

    /**
     * 把验证码送达手机号。
     *
     * @return true 送达成功；false 通道未接入/被显式禁用，调用方应回退到「请联系管理员」
     * @throws RuntimeException 通道异常（如阿里云返回错误），由全局异常拦截转 5xx
     */
    boolean send(String phone, String code);

    /**
     * 通道名，用于日志和健康检查（actuator）。
     */
    String name();
}
